File: lesson_11/arrays_ts/src/lesson11.test.ts
import { describe, expect, it } from '@jest/globals';
import { findWordsContaining, getConcatenation } from './lesson11.js';

describe('Lesson11', () => {
  describe('getConcatenation', () => {
    it('should return the correct concatenation for test case 1', () => {
      const nums1 = [1, 2, 1];
      const expected1 = [1, 2, 1, 1, 2, 1];
      const result1 = getConcatenation(nums1);
      expect(result1).toEqual(expected1);
    });

    it('should return the correct concatenation for test case 2', () => {
      const nums2 = [1, 3, 2, 1];
      const expected2 = [1, 3, 2, 1, 1, 3, 2, 1];
      const result2 = getConcatenation(nums2);
      expect(result2).toEqual(expected2);
    });

    it('should return the correct concatenation for test case 3', () => {
      const nums3: number[] = [];
      const expected3: number[] = [];
      const result3 = getConcatenation(nums3);
      expect(result3).toEqual(expected3);
    });

    it('should return the correct concatenation for test case 4', () => {
      const nums4 = [5];
      const expected4 = [5, 5];
      const result4 = getConcatenation(nums4);
      expect(result4).toEqual(expected4);
    });

    it('should return the correct concatenation for test case 5', () => {
      const nums5 = [0, 0, 0];
      const expected5 = [0, 0, 0, 0, 0, 0];
      const result5 = getConcatenation(nums5);
      expect(result5).toEqual(expected5);
    });
  });

  describe('findWordsContaining', () => {
    it('should return the correct indices for test case 1', () => {
      const ch1 = 'a';
      const words1 = ['apple', 'banana', 'cherry', 'date'];
      const expected1 = [0, 1, 3];
      const result1 = findWordsContaining(words1, ch1);
      expect(result1).toEqual(expected1);
    });

    it('should return the correct indices for test case 2', () => {
      const ch2 = 'z';
      const words2 = ['apple', 'banana', 'cherry', 'date'];
      const expected2: number[] = [];
      const result2 = findWordsContaining(words2, ch2);
      expect(result2).toEqual(expected2);
    });

    it('should return the correct indices for test case 3', () => {
      const ch3 = 'e';
      const words3 = ['apple', 'banana', 'cherry', 'date'];
      const expected3 = [0, 2, 3];
      const result3 = findWordsContaining(words3, ch3);
      expect(result3).toEqual(expected3);
    });

    it('should return the correct indices for test case 4', () => {
      const ch4 = 'a';
      const words4 = ['', ' ', 'banana'];
      const expected4 = [2];
      const result4 = findWordsContaining(words4, ch4);
      expect(result4).toEqual(expected4);
    });
  });
});

File: lesson_11/arrays_java/arrays_app/src/test/java/com/codedifferently/lesson11/Lesson11Test.java
package com.codedifferently.lesson11;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.Arrays;
import java.util.List;
import org.junit.jupiter.api.Test;

class Lesson11Test {

  @Test
  public void testGetConcatenation() {
    Lesson11 solution = new Lesson11();

    // Test case 1
    int[] nums1 = {1, 2, 1};
    int[] expected1 = {1, 2, 1, 1, 2, 1};
    int[] result1 = solution.getConcatenation(nums1);
    assertArrayEquals(expected1, result1);

    // Test case 2
    int[] nums2 = {1, 3, 2, 1};
    int[] expected2 = {1, 3, 2, 1, 1, 3, 2, 1};
    int[] result2 = solution.getConcatenation(nums2);
    assertArrayEquals(expected2, result2);

    // Test case 3
    int[] nums3 = {};
    int[] expected3 = {};
    int[] result3 = solution.getConcatenation(nums3);
    assertArrayEquals(expected3, result3);

    // Test case 4
    int[] nums4 = {5};
    int[] expected4 = {5, 5};
    int[] result4 = solution.getConcatenation(nums4);
    assertArrayEquals(expected4, result4);

    // Test case 5
    int[] nums5 = {0, 0, 0};
    int[] expected5 = {0, 0, 0, 0, 0, 0};
    int[] result5 = solution.getConcatenation(nums5);
    assertArrayEquals(expected5, result5);
  }

  @Test
  public void testWordsContainingChar() {
    Lesson11 solution = new Lesson11();

    // Test case 1
    char ch1 = 'a';
    String[] words1 = {"apple", "banana", "cherry", "date"};
    List<Integer> expected1 = Arrays.asList(0, 1, 3);
    List<Integer> result1 = solution.findWordsContaining(words1, ch1);
    assertEquals(expected1, result1);

    // Test case 2
    char ch2 = 'z';
    String[] words2 = {"apple", "banana", "cherry", "date"};
    List<Integer> expected2 = Arrays.asList();
    List<Integer> result2 = solution.findWordsContaining(words2, ch2);
    assertEquals(expected2, result2);

    // Test case 3
    char ch3 = 'e';
    String[] words3 = {"apple", "banana", "cherry", "date"};
    List<Integer> expected3 = Arrays.asList(0, 2, 3);
    List<Integer> result3 = solution.findWordsContaining(words3, ch3);
    assertEquals(expected3, result3);

    // Test case 4
    char ch4 = 'a';
    String[] words4 = {"", " ", "banana"};
    List<Integer> expected4 = Arrays.asList(2);
    List<Integer> result4 = solution.findWordsContaining(words4, ch4);
    assertEquals(expected4, result4);
  }
}
